File: apps/reading/src/api/floatingResearchDraftCombinedDocument.ts
/**
 * Floating research → provisional combined draft document (pure).
 *
 * Operator vision: after floating deep research, create a draft version with
 * the combined document (parent + research findings) before fully merging.
 *
 * draft_written always false — provisional scaffold only; never mutates parent.
 * merge_executed always false.
 */

export type DraftSourceStatus = "proposed" | "open" | "completed" | "closed";

export interface FloatingDraftSource {
  instance_id: string;
  parent_asset_id: string;
  status: DraftSourceStatus;
  highlight?: string;
  /** Caller-supplied findings / HTML fragments — never invented. */
  findings?: string[];
}

export interface ProvisionalCombinedDraft {
  parent_asset_id: string;
  instance_ids: string[];
  /** Provisional HTML-ish sections for preview (caller-supplied content only). */
  sections: string[];
  section_count: number;
  /** True when ≥1 finding or highlight section present. */
  draft_ready: boolean;
  operator_ack: boolean;
  /** Always false — pure layer never writes draft asset. */
  draft_written: false;
  /** Always false — pure layer never merges into parent. */
  merge_executed: false;
  notes: string[];
  authority: "floating_research_draft_combined_document_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Build a provisional combined draft document from parent excerpt + floating
 * research sources. Never invents findings; never writes assets.
 */
export function composeFloatingResearchDraftCombinedDocument(input: {
  parent_asset_id: string;
  /** Optional parent document excerpt/HTML (caller-supplied). */
  parent_excerpt?: string | null;
  sources: FloatingDraftSource[];
  operator_ack: boolean;
}): ProvisionalCombinedDraft {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const parent_asset_id = requireNonEmpty(
    input.parent_asset_id,
    "parent_asset_id",
  );
  if (!Array.isArray(input.sources) || input.sources.length === 0) {
    throw new Error("sources must be a non-empty array");
  }

  const notes: string[] = [
    "draft_written=false — provisional combined document not persisted",
    "merge_executed=false — parent asset not mutated",
    "section content is caller-supplied only (no invent)",
  ];

  const sections: string[] = [];
  if (input.parent_excerpt != null && input.parent_excerpt !== undefined) {
    if (typeof input.parent_excerpt !== "string" || !input.parent_excerpt.trim()) {
      throw new Error("parent_excerpt must be non-empty string when set");
    }
    sections.push(
      `<section data-role="parent" data-asset="${parent_asset_id}">${input.parent_excerpt.trim()}</section>`,
    );
  } else {
    notes.push("parent_excerpt absent — draft scaffold without parent body");
  }

  const instance_ids: string[] = [];
  const seen = new Set<string>();

  for (let i = 0; i < input.sources.length; i++) {
    const s = input.sources[i];
    if (!s || typeof s !== "object") {
      throw new Error(`sources[${i}] must be an object`);
    }
    const id = requireNonEmpty(s.instance_id, `sources[${i}].instance_id`);
    const p = requireNonEmpty(
      s.parent_asset_id,
      `sources[${i}].parent_asset_id`,
    );
    if (p !== parent_asset_id) {
      throw new Error("all sources must share parent_asset_id");
    }
    if (
      s.status !== "proposed" &&
      s.status !== "open" &&
      s.status !== "completed"
    ) {
      throw new Error(
        `sources[${i}] status must be proposed|open|completed (not closed)`,
      );
    }
    if (seen.has(id)) {
      throw new Error(`duplicate instance_id: ${id}`);
    }
    seen.add(id);
    instance_ids.push(id);

    if (s.highlight != null) {
      if (typeof s.highlight !== "string" || !s.highlight.trim()) {
        throw new Error(
          `sources[${i}].highlight must be non-empty string when set`,
        );
      }
      sections.push(
        `<section data-role="highlight" data-instance="${id}">${s.highlight.trim()}</section>`,
      );
    }
    if (s.findings != null) {
      if (!Array.isArray(s.findings)) {
        throw new Error(`sources[${i}].findings must be string[] when set`);
      }
      for (let j = 0; j < s.findings.length; j++) {
        const f = s.findings[j];
        if (typeof f !== "string" || !f.trim()) {
          throw new Error(
            `sources[${i}].findings[${j}] must be non-empty string`,
          );
        }
        sections.push(
          `<section data-role="finding" data-instance="${id}">${f.trim()}</section>`,
        );
      }
    }
  }

  const hasResearchContent = sections.some(
    (sec) =>
      sec.includes('data-role="highlight"') ||
      sec.includes('data-role="finding"'),
  );
  const draft_ready = hasResearchContent;
  if (!draft_ready) {
    notes.push(
      "draft_ready=false — no highlight/finding content (no invent sections)",
    );
  } else {
    notes.push(
      `draft_ready=true · sections=${sections.length} · instances=${instance_ids.length}`,
    );
  }
  if (!input.operator_ack) {
    notes.push(
      "operator_ack=false — preview-only; still draft_written=false",
    );
  } else {
    notes.push(
      "operator_ack=true — ack for draft preview only; still draft_written=false",
    );
  }
  notes.push("draft_written=false");
  notes.push("merge_executed=false");

  return {
    parent_asset_id,
    instance_ids,
    sections,
    section_count: sections.length,
    draft_ready,
    operator_ack: input.operator_ack,
    draft_written: false,
    merge_executed: false,
    notes,
    authority: "floating_research_draft_combined_document_advisory",
  };
}

export function formatFloatingDraftCombinedSummary(
  d: ProvisionalCombinedDraft,
): string {
  return (
    `draft combined · sections=${d.section_count} · ready=${d.draft_ready} · ` +
    `draft_written=false · merge_executed=false`
  );
}

File: substrate/floating_research_draft_combined_document.py
"""Floating research provisional combined draft document (pure).

Parent excerpt + floating findings → provisional draft sections.
draft_written and merge_executed always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


class FloatingResearchDraftCombinedDocumentError(ValueError):
    """Fail-closed validation for provisional combined draft."""


@dataclass(frozen=True)
class ProvisionalCombinedDraft:
    parent_asset_id: str
    instance_ids: tuple[str, ...]
    sections: tuple[str, ...]
    section_count: int
    draft_ready: bool
    operator_ack: bool
    draft_written: bool
    merge_executed: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "parent_asset_id": self.parent_asset_id,
            "instance_ids": list(self.instance_ids),
            "sections": list(self.sections),
            "section_count": self.section_count,
            "draft_ready": self.draft_ready,
            "operator_ack": self.operator_ack,
            "draft_written": False,
            "merge_executed": False,
            "notes": list(self.notes),
            "authority": "floating_research_draft_combined_document_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise FloatingResearchDraftCombinedDocumentError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_floating_research_draft_combined_document(
    *,
    parent_asset_id: object,
    sources: object,
    operator_ack: object,
    parent_excerpt: object | None = None,
) -> ProvisionalCombinedDraft:
    """Build provisional combined draft. Never writes assets."""
    if not isinstance(operator_ack, bool):
        raise FloatingResearchDraftCombinedDocumentError(
            "operator_ack must be an explicit boolean"
        )
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")
    if not isinstance(sources, list) or len(sources) == 0:
        raise FloatingResearchDraftCombinedDocumentError(
            "sources must be a non-empty array"
        )

    notes: list[str] = [
        "draft_written=false — provisional combined document not persisted",
        "merge_executed=false — parent asset not mutated",
        "section content is caller-supplied only (no invent)",
    ]
    sections: list[str] = []
    if parent_excerpt is not None:
        if not isinstance(parent_excerpt, str) or not parent_excerpt.strip():
            raise FloatingResearchDraftCombinedDocumentError(
                "parent_excerpt must be non-empty string when set"
            )
        sections.append(
            f'<section data-role="parent" data-asset="{parent}">'
            f"{parent_excerpt.strip()}</section>"
        )
    else:
        notes.append("parent_excerpt absent — draft scaffold without parent body")

    instance_ids: list[str] = []
    seen: set[str] = set()
    for i, s in enumerate(sources):
        if not isinstance(s, dict):
            raise FloatingResearchDraftCombinedDocumentError(
                f"sources[{i}] must be an object"
            )
        iid = _require_nonempty(
            s.get("instance_id"), field=f"sources[{i}].instance_id"
        )
        p = _require_nonempty(
            s.get("parent_asset_id"), field=f"sources[{i}].parent_asset_id"
        )
        if p != parent:
            raise FloatingResearchDraftCombinedDocumentError(
                "all sources must share parent_asset_id"
            )
        status = s.get("status")
        if status not in ("proposed", "open", "completed"):
            raise FloatingResearchDraftCombinedDocumentError(
                f"sources[{i}] status must be proposed|open|completed (not closed)"
            )
        if iid in seen:
            raise FloatingResearchDraftCombinedDocumentError(
                f"duplicate instance_id: {iid}"
            )
        seen.add(iid)
        instance_ids.append(iid)

        highlight = s.get("highlight")
        if highlight is not None:
            if not isinstance(highlight, str) or not highlight.strip():
                raise FloatingResearchDraftCombinedDocumentError(
                    f"sources[{i}].highlight must be non-empty string when set"
                )
            sections.append(
                f'<section data-role="highlight" data-instance="{iid}">'
                f"{highlight.strip()}</section>"
            )
        findings = s.get("findings")
        if findings is not None:
            if not isinstance(findings, list):
                raise FloatingResearchDraftCombinedDocumentError(
                    f"sources[{i}].findings must be string[] when set"
                )
            for j, f in enumerate(findings):
                if not isinstance(f, str) or not f.strip():
                    raise FloatingResearchDraftCombinedDocumentError(
                        f"sources[{i}].findings[{j}] must be non-empty string"
                    )
                sections.append(
                    f'<section data-role="finding" data-instance="{iid}">'
                    f"{f.strip()}</section>"
                )

    has_research = any(
        'data-role="highlight"' in sec or 'data-role="finding"' in sec
        for sec in sections
    )
    draft_ready = has_research
    if not draft_ready:
        notes.append(
            "draft_ready=false — no highlight/finding content (no invent sections)"
        )
    else:
        notes.append(
            f"draft_ready=true · sections={len(sections)} · "
            f"instances={len(instance_ids)}"
        )
    if not operator_ack:
        notes.append(
            "operator_ack=false — preview-only; still draft_written=false"
        )
    else:
        notes.append(
            "operator_ack=true — ack for draft preview only; still draft_written=false"
        )
    notes.append("draft_written=false")
    notes.append("merge_executed=false")

    return ProvisionalCombinedDraft(
        parent_asset_id=parent,
        instance_ids=tuple(instance_ids),
        sections=tuple(sections),
        section_count=len(sections),
        draft_ready=draft_ready,
        operator_ack=operator_ack,
        draft_written=False,
        merge_executed=False,
        notes=tuple(notes),
        authority="floating_research_draft_combined_document_advisory",
    )


__all__ = [
    "FloatingResearchDraftCombinedDocumentError",
    "ProvisionalCombinedDraft",
    "compose_floating_research_draft_combined_document",
]
